staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -539,11 +539,17 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 
 	// determine the namespace and name scope of the watch, first from the request, secondarily from the field selector
 	scope := namespacedName{}
-	if requestNamespace, ok := request.NamespaceFrom(ctx); ok && len(requestNamespace) > 0 {
-		scope.namespace = requestNamespace
-	} else if selectorNamespace, ok := pred.Field.RequiresExactMatch("metadata.namespace"); ok {
-		scope.namespace = selectorNamespace
+
+	// for request like '/api/v1/watch/namespaces/*', don't set scope.namespace.
+	// As the function NamespaceToSelectableFields don't set the filed metadata.namespace.
+	if c.groupResource.Resource != "namespaces" {
```

Comment:
if we compare to a special case resource, this has to compare to a fully-qualified resource (group="", resource="namespaces")

I would make this special case a separate case after scope.name and scope.namespace are selected:

```go
var coreNamespace = schema.GroupResource{Group: "", Resource: "namespaces"}
...
// special-case watch of a single namespace to work properly via /api/v1/watch/namespaces/*
// namespaces don't populate metadata.namespace in ObjFields.
if c.groupResource == coreNamespace && len(scope.namespace) > 0 && scope.namespace == scope.name {
  scope.namespace = ""
}
```

Reply:
alternatively, in `startDispatching`, we could populate `namespace` to match `name` for coreNamespace objects only (but this requires evaluating logic on every event, which seems less preferable to evaluating a special case once at watch start here)

Reply:
Got it.